Initial ParameterTransformer support

JamesNK opened the pull request:
- [x] Attribute routes token replacement
- [x] MvcEndpointDataSource endpoint parameter replacement
- [x] Fix endpoint support for area/controller/action in attribute route
- [x] Unit test matching with parameter transformer
- [x] Unit test link generation with parameter transformer
- [x] XML docs

JamesNK commented:
I was thinking about attribute routing token replacement and how to make a parameter transformer work. I can think of two options:
1. Extend the token replacement syntax to support parameter policies. This feels like overkill to me just for declaring a parameter transformer. Would another policy like a constraint make sense in an attribute route?
2. Add `ParameterTransformer MvcOptions.AttributeRoutingParameterTransformer { get; set; }` (better name welcome) that is used to resolve the values of attribute route tokens. The downside of this is it is a big global hammer - it runs on all attribute routes and all tokens. Also, would there be tokens other than area/controller/action that the dev wouldn't want transformed?

Thoughts?

JamesNK commented:
Some more thoughts about combining parameter transformer with attribute routing:

Today attribute routing token replacement completely removes the token from the final pattern, e.g.

Template: `[controller]/[action]` (with kebab case transformer)
Route values: Controller=Store, Action=ProductList
Final template: `store/product-list`

There are no parameters in the final template and the presence of the transformer has been erased.

If we are generating a URL for this endpoint with the same route values, we would want `store/product-list` as the output. Would URL generation today create this?

rynowak commented:
Response to come of your comments inline (I haven't looked at the PR yet  😆)

> Extend the token replacement syntax to support inline parameter policies. This feels like overkill to me just for declaring a parameter transformer. Would another policy like a constraint make sense in an attribute route?

Yeah that really seems like a lot of work for something that's kind of a dead end. 

We could consider also allowing `{action}` and friends in attribute routes - where they traditionally have not been allowed (hear me out). I think the tricky part is that we'd only support this for endpoint routing if at all? 

By now the additional effort of getting folks to understand the difference between `{ }` and `[ ]` is done. We created the `[ ]` to make it clear that `{ }` is for 'runtime' parameters, and `[ ]` is for static replacement. (If I were a superhero my name would be **Mr Syntactic Salt** and my super power would be making sure the semantic details of everything are obvious to all by the most oblivious user. @pranavkm on the other hand would just be **Mr Salt** but I digress....)

Now we have to support `{ }` for static replacement due to conventional routing and it has **MORE** features that `[ ]` does. We also have https://github.com/aspnet/Routing/issues/557 which is only my 'most wanted' features dating back to like beta5 of DNX/ASP.NET 5 (yes before ASP.NET Core was adopted).

I'm interested in your thoughts about this one....

> Add ParameterTransformer AttributeRouteTokenTransformer { get; set; } (better name welcome) to MvcOptions that is used to resolve the values of attribute route tokens. The downside of this is it is a big global hammer - it runs on all attribute routes and all tokens. Also, would there be tokens other than area/controller/action that the dev wouldn't want transformed?

I thought you wanted to do this anyway 😆 I remember that you were advocating for a feature that would turn on a parameter transformer for all of the 'mvc' route keys. I'm not opposed to this - I agree that the naming and design will be a challenge. Maybe its ok if it's not appealing to use. 

> Some more thoughts about combining parameter transformer with attribute routing:

> Today attribute routing token replacement completely removes the token from the final pattern, e.g.
> Template: [controller]/[action] (with kebab case transformer)
> Route values: Controller=Store, Action=ProductList
> Final template: store/product-list

> There are no parameters in the final template and the presence of the transformer has been erased.

> If we are generating a URL for this endpoint with the same route values, we would want store/product-list as the output. 

Yes, I agree, parameter transformers need to operate by-directionally. 

Logically - the value `product-list` should be decoded to `ProductList` for URL matching resulting in `{... action = "ProductList" }`. Likewise the route value `{... action = "ProductList" }` should be encoded when doing URL generation to `product-list`.  This is a logical transformation - the implementation will do this statically when building the endpoint. They key is that the route value for `action` is always **just** the name of the method. Keep in mind that route values appear hardcoded in a lot of places.



> Would URL generation today create this?

I assume you mean "Would conventional URL generation today create this?". The answer is the conventional URL generation will use the text you give it verbatim, which is why a transformation like lowercasing works OK, but a tranformation like kebab-casing doesn't. 

I'm not worried about compatibility for cases where someone hardcoded their preferred style of values, because we're giving that user a new system with a more useful feature. They can revert to 2.1 mode to use something exactly compatible.

JamesNK commented:
🆙 📅 

* Support setting a parameter transformer against route attribute model
* Fix area/controller/action parameters in route attribute with endpoint routing
* Support parameter transformer in route attribute

JamesNK commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ParameterTransformerConvention.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Microsoft.AspNetCore.Mvc.Core.ApplicationModels
+{
+    public sealed class RouteTokenTransformerConvention : IApplicationModelConvention
```
This doesn't need to come with MVC. Can be something that people who want to use a transformer could include with there app

rynowak replied:
I think this is pretty nice.

JamesNK commented on test/Microsoft.AspNetCore.Mvc.FunctionalTests/RoutingTestsBase.cs:
```diff
@@ -36,6 +36,21 @@ protected RoutingTestsBase(MvcTestFixture<TStartup> fixture)
         public abstract Task RouteData_Routers_AttributeRoute();
 
+        //[Fact]
+        //public async Task sdfsdf()
```
Derp

JamesNK commented on test/WebSites/RoutingWebSite/Startup.cs:
```diff
@@ -14,8 +15,19 @@ public class Startup
+            services
+                .AddMvc(options =>
+                {
+                    options.Conventions.Add(new ControllerRouteTokenTransformerConvention(
```
Only want to test parameter transformer with one controller in tests

JamesNK commented on test/WebSites/RoutingWebSite/StartupWith21Compat.cs:
```diff
@@ -14,11 +19,16 @@ public class StartupWith21Compat
             services.AddScoped<TestResponseGenerator>();
             services.AddSingleton<IActionContextAccessor, ActionContextAccessor>();
+
+            var actionDescriptorProvider = new RemoveControllerActionDescriptorProvider(typeof(EndpointRoutingController));
```
Old routing doesn't support parameters in attributes so remove its actions to avoid exception

JamesNK commented:
TODO: Link generation tests to parameter transformed endpoints in MVC

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/AttributeRouteModel.cs:
```diff
@@ -66,6 +68,8 @@ public AttributeRouteModel(AttributeRouteModel other)
 
         public bool IsAbsoluteTemplate => Template != null && IsOverridePattern(Template);
 
+        public IParameterTransformer RouteTokenTransformer { get; set; }
```
Just wanted to point out that this API shape implies a single parameter transformer for all of the 'mvc' parameters. This is probably the most common usage, but the system is intended to be much more flexible.

rynowak replied:
This design makes sense though based on the design of the `ReplaceTokens` methods.

JamesNK replied:
This value is per attribute route model. There is nothing stopping someone assigning different transformers on different attribute route model. The unit tests already do this (only applies a transformer to a single controller for testing purposes)

rynowak replied:
Right but it's not per-parameter.

JamesNK replied:
oic, yes correct

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ParameterTransformerConvention.cs:
```diff
@@ -0,0 +1,38 @@
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Mvc.Core.ApplicationModels
```
drop `.Core`

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ParameterTransformerConvention.cs:
```diff
@@ -0,0 +1,38 @@
+    public sealed class RouteTokenTransformerConvention : IApplicationModelConvention
+    {
+        private readonly IParameterTransformer _parameterTransformer;
+
```
Consider implementing `IActionModelConvention`, unsealing, and adding a `protected virtual bool ShouldApply(...)` method.  Similar to https://github.com/aspnet/Mvc/pull/8390

rynowak replied:
This lets users subclass the convention to apply it or not apply it based on some criteria they define.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Abstractions/Routing/AttributeRouteInfo.cs:
```diff
@@ -36,5 +38,7 @@ public class AttributeRouteInfo
         /// </summary>
         public bool SuppressPathMatching { get; set; }
+
+        public IParameterTransformer RouteTokenTransformer { get; set; }
```
So one issue that comes to mind here is that this will keep all of the transformer instance around in memory long after they have served their purpose, that's really not ideal.  

Do we really need this here or just on the `AttributeRouteModel`?

JamesNK replied:
Yeah it needs to be here because `AttributeRouteModel.ReplaceTokens` uses the values from `AttributeRouteInfo`.

Could set it to null after token replacement has run? And maybe also make it internal on AttributeRouteInfo?

rynowak replied:
That's a good stop gap, but why couldn't this be passed as parameter to `ReplaceTokens`? What else would need to change?

JamesNK replied:
It is called from `ControllerActionDescriptorBuilder.Build`. It need to come somewhere from `ApplicationModel`.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs:
```diff
@@ -391,13 +393,15 @@ private static AttributeRouteInfo CreateAttributeRouteInfo(AttributeRouteModel r
                         actionDescriptor.AttributeRouteInfo.Name,
-                        actionDescriptor.RouteValues);
+                        actionDescriptor.RouteValues,
+                        actionDescriptor.AttributeRouteInfo.RouteTokenTransformer);
```
Possible to pass this into this method rather than using `ARI` to truck it around?

JamesNK replied:
No, it has to come from `ApplicationModel`. Maybe it could live in `ControllerActionDescriptor.Properties`?

rynowak replied:
Well `ApplicationModel ` is fine. `AttributeRouteInfo` is less fine

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs:
```diff
@@ -14,6 +14,7 @@
 using Microsoft.AspNetCore.Mvc.ModelBinding.Binders;
 using Microsoft.AspNetCore.Mvc.ModelBinding.Metadata;
 using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
+using Microsoft.AspNetCore.Routing;
```
???

JamesNK commented:
🆙 📅

JamesNK commented:
🆙 📅 

Updated to remove the transformer from `AttributeRouteInfo`.

I also removed it from `AttributeRouteModel` since it and the info class currently mirror each other, and using a property throughout the entire process worked well.

JamesNK commented:
FYI I'm going to split out the fix for replacing route attribute parameters into its own PR, merge that, and rebase this on top. Overall changes will be the same.

rynowak commented:
Alright, ping me if there's anything you need to me to look at
